Implementation of Parallelization to `MDAnalysis.analysis.contacts`

talagayev opened the pull request:
Fixes #4660 

Changes made in this Pull Request:
 - added backends and aggregators to `Contacts` in `analysis.contacts`
 - added `get_box` function in `analysis.contacts` to fix error, which appeared during parallelization
 - added the `client_Contacts` in `conftest.py`
 - added `client_Contacts` in `run()` in `test_contacts.py`

The function `get_box` was implemented due to this Error occuring in pytests with parallelization:
`AttributeError: Can't get local object 'Contacts.__init__.<locals>.<lambda>'`

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4820.org.readthedocs.build/en/4820/

<!-- readthedocs-preview mdanalysis end -->

talagayev committed:
Update contacts.py

added aggregator and supported backends + get_box to fix an error with lambda that appeared for multiprocessing & dask

talagayev committed:
Update contacts.py

black formatting

talagayev committed:
Update conftest.py

addition of client_Contacts

talagayev committed:
Update test_contacts.py

Addition of client_Contacts to pytests

talagayev committed:
Update CHANGELOG

Changelog addition for contacts parallelization

pep8speaks commented:
Hello @talagayev! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2024-12-15 00:49:24 UTC

talagayev committed:
Update contacts.py

added versionchange

talagayev committed:
Update contacts.py

PEP Fixes

talagayev committed:
Update test_contacts.py

pep fixes

talagayev committed:
Update contacts.py

import ResultsGroup

talagayev committed:
Merge branch 'develop' into parallel_contacts

talagayev committed:
Update CHANGELOG

moved space

RMeli commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -338,6 +338,28 @@ def contact_matrix(d, radius, out=None):
     return out
 
 
+def get_box(ts, pbc):
+    """Retrieve the dimensions of the simulation box based on PBC.
```
This function is very general, so it should be in a core module I think. Otherwise it could be a lambda defined below (see comment).

RMeli commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -445,10 +481,8 @@ def __init__(self, u, select, refgroup, method="hard_cut", radius=4.5,
-        else:
-            self._get_box = lambda ts: None
+        self.pbc = pbc
+        self._get_box = functools.partial(get_box, pbc=self.pbc)
```
```suggestion
        self._get_box = lambda ts: ts.dimensions if self.pbc else None
```
?

talagayev replied:
@RMeli I tried running it with the suggestion locally and I get

`AttributeError: Can't get local object 'Contacts.__init__.<locals>.<lambda>'`

It is somehow due to the parallelization having problems with lamba, it also does not like this


```
        self._get_box = functools.partial(
            lambda ts, pbc: ts.dimensions if pbc else None, pbc=self.pbc
        )
```
it leads to the same error, so yes I would agree that the function is quite general, so moving 
`def get_box()` to be a function in the `Contacts` class would be fine? since that would
works with the tests

orbeckst replied:
I wouldn't put `get_box()` elsewhere because it's really a bit clunky and not really that useful except that it's needed here (or do we have other places where we need it??). I'd keep it as local and non-public as possible. Writing it as a lambda would be the nicest way but if that does not work (... good to know and a bit sad ...) then I'd make it a static private class method
```python
@staticmethod
def _get_box_func(ts, pbc):
    """... keep the written docs ... """
    return ts.dimensions if pbc else None
```
which you can then use for the partial.

Add a comment to `self._get_box = functools.partial(self._get_box_func, pbc=self.pbc)` why all of this is necessary.

----

(I assume that following will run into the same issue as lambdas:
```python
def get_box(ts, pbc=self.pbc):
   return ts.dimensions if pbc else None
self._get_box = get_box
```
as it's also local, right?)

RMeli replied:
Thanks @talagayev for the explanation, I didn't know a lambda would not work in this case. I think defining a function instead of a lambda will in our in the same issue, as you suspect @orbeckst, but haven't tried. If we don't want to move the function elsewhere, I agree it should be private and not exposed by the module.

talagayev replied:
> (I assume that following will run into the same issue as lambdas:
> 
> ```python
> def get_box(ts, pbc=self.pbc):
>    return ts.dimensions if pbc else None
> self._get_box = get_box
> ```
> 
> as it's also local, right?)


Yes, tried it right now locally, leads sadly to the `Can't get local object 'Contacts.__init__.<locals>.get_box'`

talagayev replied:
but yes making it more local static method is a more sophisticated way then it is currently, will adjust it then :)

as for it being used anywhere else, not really, it is only used here once

talagayev replied:
@orbeckst @RMeli I moved it now as you suggested and would re request the review, to see if it is correct now :)

talagayev committed:
Merge branch 'develop' into parallel_contacts

talagayev committed:
Merge branch 'develop' into parallel_contacts

talagayev committed:
Update contacts.py

adjusted fixes, moved get_box into class and made
it more local and less accessable

talagayev committed:
Update contacts.py

intdent

talagayev committed:
Update contacts.py

forgot to add staticmethod, probably reason of failure in tests added it now

talagayev committed:
Update contacts.py

pep

talagayev committed:
Update conftest.py

pep fix

talagayev committed:
Update contacts.py

orbeckst reviewed (approved):
Nice work @talagayev , thank you!

talagayev commented:
> Nice work @talagayev , thank you!

Nice, happy to help :)
